### PR TITLE
RHPAM-2708 : RHPAM 7.x Business Central latency correlated with number of groups memberships

### PR DESCRIPTION
**Pull request by RishiRajAnand**
- fix the issue with caching of permissions for resources. Now, after using a resource (Screens, Spaces etc), the permissions corresponding to that resource will be cached and subsequent use will be faster.
- Optimized fetching of groups and roles permissions. Replaced HashSet based collection with HashMap to provide faster retrival of roles/group permissions.
- Please note the initial home page load might still be little slower when we use elevated number of groups since it calculates and merges permissions of all the groups. however, subsequent page load should be faster since it fetches from the permission cache

### CODE REVIEW

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

File: uberfire-security/uberfire-security-api/src/main/java/org/uberfire/security/impl/authz/DefaultAuthorizationPolicy.java [hunk 1]
```diff
@@ -30,32 +31,34 @@
 public class DefaultAuthorizationPolicy implements AuthorizationPolicy {
 
     DefaultAuthorizationEntry defaultEntry = new DefaultAuthorizationEntry();
-    private Set<DefaultAuthorizationEntry> entrySet = new HashSet<>();
+    final Map<Group, DefaultAuthorizationEntry> groupEntryMap = new HashMap<>();
+    final Map<Role, DefaultAuthorizationEntry> rolesEntryMap = new HashMap<>();
```
Comment (caponetto) [inline uberfire-security/uberfire-security-api/src/main/java/org/uberfire/security/impl/authz/DefaultAuthorizationPolicy.java]:
```suggestion
    private final Map<Group, DefaultAuthorizationEntry> groupEntryMap = new HashMap<>();
    private final Map<Role, DefaultAuthorizationEntry> rolesEntryMap = new HashMap<>();
```

Comment (RishiRajAnand) [reply uberfire-security/uberfire-security-api/src/main/java/org/uberfire/security/impl/authz/DefaultAuthorizationPolicy.java]: I have added test case covering cache functionality and recalculation of the permissions . Thanks @caponetto for your review and help this. Much appreciated :)

**Comment by RishiRajAnand** [comment]
> Hi, Rishiraj thanks for this.
> 
> Could you please add some unit tests? A good starting point would be covering "verifying the group/role is not re-calculated"

I added the test cases for cache functionality and that the permissions are fetched from cache for subsequent access.

**Comment by caponetto** [comment]
Jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
Hi @ederign @caponetto , I also add cache for permission collection object which avoid recalculating and merging group and roles permission from the policy on every resource load. Now only business-central home page load should be relatively slow and subsequent resources access should be faster.

**Comment by ederign** [comment]
Jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by ederign** [comment]
@RishiRajAnand some tests are failing on other PR's, are those failures related?

**Comment by RishiRajAnand** [comment]
Failures doesn't seem related to this but it seems the selenium test case failure in DMN designer might be related to https://issues.redhat.com/browse/DROOLS-5170 ? We didn't get these failures when we actually merge the [PR](https://github.com/kiegroup/appformer/pull/901) though.

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by ederign** [comment]
@RishiRajAnand  are the failures related?

**Comment by RishiRajAnand** [comment]
@ederign The failures are not related

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by caponetto** [comment]
@ederign Green!

**Comment by ederign** [comment]
great work @RishiRajAnand

**Pull request merged**